### PR TITLE
Add Metrics impl based on `merni`

### PR DESCRIPTION
**Pull request by Swatinem**
I pretty much copy-pasted the core `merni` code, which is not that much.

Also built shims to statsdproxy, and updated all usages of the previous `Metrics` trait and `get_metrics` to use the global macros.

### CODE REVIEW

File: rust_snuba/rust_arroyo/src/metrics/globals.rs [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+where
+    F: FnOnce(&dyn Recorder) -> R,
+    R: Default,
+{
+    if let Some(recorder) = GLOBAL_RECORDER.get() {
+        f(recorder)
+    } else {
+        Default::default()
+    }
+}
```
Comment (loewenheim) [inline rust_snuba/rust_arroyo/src/metrics/globals.rs:54]: Is there a reason this shouldn't just be one function?
```suggestion
pub fn record_metric(metric: Metric<'_>) {
    if let Some(recorder) = GLOBAL_RECORDER.get() {
        recorder.record_metric(metric)
    }
}
```

File: rust_snuba/src/metrics/statsd.rs [hunk 2]
```diff
@@ -1,15 +1,28 @@
...
+
+impl Recorder for StatsDBackend {
+    fn record_metric(&self, metric: Metric<'_>) {
+        self.recorder.record_metric(metric)
+    }
+}
+
+struct Wrapper(Box<dyn cadence::MetricSink + Send + Sync + 'static>);
+
+impl MetricSink for Wrapper {
```
Comment (loewenheim) [inline rust_snuba/src/metrics/statsd.rs:22]: Couldn't this be implemented directly on the `Box<dyn …>`?

Comment (Swatinem) [reply rust_snuba/src/metrics/statsd.rs:22]: Not here at least, but rather where `MetricSink` is implemented due to orphan rules.

Comment (loewenheim) [reply rust_snuba/src/metrics/statsd.rs:22]: Oh, sorry, I missed that this wasn't happening in `rust_arroyo`. Should the `cadence` compatibility stuff perhaps be moved there?

Comment (loewenheim) [reply rust_snuba/src/metrics/statsd.rs:22]: Sounds reasonable

**Comment by untitaker** [comment]
merging this for now. two things to follow up with:

* merge merni into statsdproxy IMO (open to alternative suggestions but it makes sense to me at least)
* get rid of cadence entirely, right now we still use it

**Pull request merged**